src/rt/sched/behaviourcore.h

```diff
@@ -569,41 +951,174 @@ namespace verona::rt
 
   inline void Slot::release()
   {
-    assert(!is_wait());
+    Logging::cout() << "Release slot " << *this << Logging::endl;
+
+    assert(!is_wait_2pl());
 
     // This slot represents a duplicate cown, so we can ignore releasing it.
-    if (cown == nullptr)
+    if (cown() == nullptr)
+    {
+      Logging::cout() << "Duplicate cown slot " << *this << Logging::endl;
       return;
+    }
 
-    if (is_ready())
+    if (is_read_only())
     {
-      yield();
-      auto slot_addr = this;
-      // Attempt to CAS cown to null.
-      if (cown->last_slot.compare_exchange_strong(
-            slot_addr, nullptr, std::memory_order_acq_rel))
+      if (next_slot() == nullptr)
       {
-        yield();
-        Logging::cout() << "No more work for cown " << cown << Logging::endl;
-        // Success, no successor, release scheduler threads reference count.
-        shared::release(ThreadAlloc::get(), cown);
-        return;
+        auto slot_addr = this;
+        if (cown()->last_slot.compare_exchange_strong(
+              slot_addr, nullptr, std::memory_order_acq_rel))
+        {
+          Logging::cout() << *this << " Reader releasing the cown "
+                          << Logging::endl;
```

Comment:
I have a similar comment that the skeleton of the read and write cases are pretty similar. I would prefer the `is_read_only` to be pushed inside the checks further to remove code duplication.

```C++
if (next_slot() == nullptr)
{
   auto slot_addr = this;
   if (cown()->last_slot.compare_exchange_strong(
       slot_addr, nullptr, std::memory_order_acq_rel))
   {
      if (is_read_only() && cown()->read_ref_count.release_read())
      {
         // Attempt to wake next writer
         ...
         shared::release(ThreadAlloc::get(), cown());
      }

      yield();
      shared::release(ThreadAlloc::get(), cown());
      return;
   }

   while (next_slot() == nullptr)
   {
     Systematic::yield_until(
       [this]() { return (next_slot() != nullptr); });
     Aal::pause();
   }
}

if (is_read_only())
{
  ...
  return;
}

...
```